### PR TITLE
Enhancement/convert firebase functions to ts

### PR DESCRIPTION
**Pull request by jagaSto**
**What changes does this PR introduce**

* converts the firebase functions from JS to TS 🚀 
* bumps node version from 8 to 10
* as of now still a draft since there are some open discussion points. 
* the build process needs to be tested on staging and we need to perform some sanity checks that everything is still working!

For references, see:
* https://firebase.google.com/docs/functions/typescript
* https://github.com/firebase/functions-samples/tree/master/typescript-getting-started

### CODE REVIEW

File: firebase/functions/src/types/collections/AskForHelpCollectionEntry.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+export interface AskForHelpCollectionEntry {
+  uid: string // TODO: check if also available outside of d
```
Comment (jagaSto) [inline firebase/functions/src/types/collections/AskForHelpCollectionEntry.ts:2]: can somebody confirm that this is indeed the case? Not visible to me from looking at the staging collection entries

File: firebase/functions/tslint.json [hunk 2]
```diff
@@ -0,0 +1,108 @@
+{
+  "rules": {
```
Comment (jagaSto) [inline firebase/functions/tslint.json:2]: default rules created with `firebase init functions`

Comment (jagaSto) [reply firebase/functions/tslint.json:2]: open for discussion I guess, but I'm pretty much fine with how they are currently!

File: firebase/functions/src/types/collections/AskForHelpCollectionEntry.ts (hunk 1 above)
Comment (kenodressel) [reply firebase/functions/src/types/collections/AskForHelpCollectionEntry.ts:2]: its not

File: firebase/functions/src/types/collections/NotificationsCollectionEntry.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+export interface NotificationsCollectionEntry {
+  d: {
+    uid: string
+    timestamp: string
```
Comment (kenodressel) [inline firebase/functions/src/types/collections/NotificationsCollectionEntry.ts:4]: should be an integer
```suggestion
    timestamp: number
```

File: firebase/functions/src/types/collections/AskForHelpCollectionEntry.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  d: {
+    uid: string
+    timestamp: string
```
Comment (kenodressel) [inline firebase/functions/src/types/collections/AskForHelpCollectionEntry.ts:5]: should be int
```suggestion
    timestamp: number
```

File: firebase/functions/src/types/collections/ReportedPostsCollectionEntry.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export interface ReportedPostsCollectionEntry {
+  uid: string
+  timestamp: string
```
Comment (kenodressel) [inline firebase/functions/src/types/collections/ReportedPostsCollectionEntry.ts:3]:
```suggestion
  timestamp: number
```

File: firebase/functions/src/types/collections/NotificationsCollectionEntry.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    email: string
+    location: string
+    plz: string
+    coordinates: FirebaseFirestore.GeoPoint
+    distance?: number // set on searchAndSendNotificationEmails
```
Comment (kenodressel) [inline firebase/functions/src/types/collections/NotificationsCollectionEntry.ts:9]: not a single entry has this property, should we remove it?

Comment (jagaSto) [reply firebase/functions/src/types/collections/NotificationsCollectionEntry.ts:9]: as mentioned in the comment, it is set as a part of `searchAndSendNotificationEmails ` to determine the closed offers for an ask-for-help entry, but never persisted in the database

**Comment by jagaSto** [comment]
tested on staging following this list https://github.com/option26/quarantine-hero/issues/335

**Pull request merged**